test_conformance/subgroups/workgroup_common_templates.h

```diff
@@ -0,0 +1,778 @@
+        case ArithmeticOp::and_: return a & b;
+        case ArithmeticOp::or_: return a | b;
+        case ArithmeticOp::xor_: return a ^ b;
+        case ArithmeticOp::logical_and: return a && b;
+        case ArithmeticOp::logical_or: return a || b;
+        case ArithmeticOp::logical_xor: return !a ^ !b;
+        default: log_error("Unknown operation request"); break;
+    }
+    return 0;
+}
+// Specialize for floating points.
+template <>
+inline cl_double calculate(cl_double a, cl_double b, ArithmeticOp operation)
+{
+    switch (operation)
+    {
+        case ArithmeticOp::add_: {
+            return a + b;
+        }
+        case ArithmeticOp::max_: {
+            return a > b ? a : b;
+        }
+        case ArithmeticOp::min_: {
+            return a < b ? a : b;
+        }
+        case ArithmeticOp::mul_: {
+            return a * b;
+        }
+        default: log_error("Unknown operation request"); break;
+    }
+    return 0;
+}
+
+template <>
+inline cl_float calculate(cl_float a, cl_float b, ArithmeticOp operation)
+{
+    switch (operation)
+    {
+        case ArithmeticOp::add_: {
+            return a + b;
+        }
+        case ArithmeticOp::max_: {
+            return a > b ? a : b;
+        }
+        case ArithmeticOp::min_: {
+            return a < b ? a : b;
+        }
+        case ArithmeticOp::mul_: {
+            return a * b;
+        }
+        default: log_error("Unknown operation request"); break;
+    }
+    return 0;
+}
+
+template <>
+inline subgroups::cl_half calculate(subgroups::cl_half a, subgroups::cl_half b,
+                                    ArithmeticOp operation)
+{
+    switch (operation)
+    {
+        case ArithmeticOp::add_: return to_half(to_float(a) + to_float(b));
+        case ArithmeticOp::max_:
+            return to_float(a) > to_float(b) || is_half_nan(b.data) ? a : b;
+        case ArithmeticOp::min_:
+            return to_float(a) < to_float(b) || is_half_nan(b.data) ? a : b;
+        case ArithmeticOp::mul_: return to_half(to_float(a) * to_float(b));
+        default: log_error("Unknown operation request"); break;
+    }
+    return to_half(0);
+}
+
+template <typename Ty> bool is_floating_point()
+{
+    return std::is_floating_point<Ty>::value
+        || std::is_same<Ty, subgroups::cl_half>::value;
+}
+
+template <typename Ty, ArithmeticOp operation>
+void genrand(Ty *x, Ty *t, cl_int *m, int ns, int nw, int ng)
+{
+    int nj = (nw + ns - 1) / ns;
+
+    for (int k = 0; k < ng; ++k)
+    {
+        for (int j = 0; j < nj; ++j)
+        {
+            int ii = j * ns;
+            int n = ii + ns > nw ? nw - ii : ns;
+
+            for (int i = 0; i < n; ++i)
+            {
+                cl_ulong x;
+                if ((operation == ArithmeticOp::add_
+                     || operation == ArithmeticOp::mul_)
+                    && is_floating_point<Ty>())
+                {
+                    // work around different results depending on operation
```

Comment:
There's a lot to keep track of here as the range-reduced but possibly sign extended integer is assigned to a cl_ulong with possible underflow. Then the lower 16 bits are assigned to a cl_half via modulo arithmetic. If I have all that right, you would end up with a distribution consisting of only NaNs (all 1s in the top bits) and positive denormals (only lowest 7 bits set).  Both are problematic for these tests. Let me know if I've missed the boat on this, but it feels like the floats will have to be constructed very carefully to ensure that they only consist of low-precision normal floats of similar magnitude and ideally both positive and negative values.